catkit2/services/oceanoptics_spectrometer_sim/oceanoptics_spectrometer_sim.py

```diff
@@ -0,0 +1,153 @@
+
+    Attributes
+    ----------
+    spectrometer : Spectrometer
+        The spectrometer to control.
+    serial_number : str
+        The serial number of the spectrometer.
+    pixels_number : int
+        Number of pixels in the spectrum.
+    interval : float
+        The interval between spectrum acquisitions in seconds.
+    spectra : DataStream
+        A data stream to submit spectra from the spectrometer.
+    is_saturating : DataStream
+        A data stream to submit whether the spectrometer is currently saturating.
+    NUM_FRAMES : int
+        The number of frames to allocate for the data streams.
+
+    Methods
+    -------
+    open()
+        Open the service.
+    main()
+        The main function of the service.
+    close()
+        Close the service.
+    take_one_spectrum()
+        Get a spectrum from the spectrometer.
+    '''
+    NUM_FRAMES = 20
+
+    def __init__(self):
+        '''
+        Create a new OceanOpticsSpectrometer service.
+        '''
+        super().__init__('oceanoptics_spectrometer_sim')
+
+        self.serial_number = str(self.config['serial_number'])
+        self.interval = self.config.get('interval', 1)
+        self._exposure_time = self.config.get('exposure_time', 1000)
+
+
+        self.pixels_number = 3000
+
+        self.spectra = None
+        self.is_saturating = None
+
+    def open(self):
+        '''
+        Open the service.
+
+        This function is called when the service is opened.
+        It initializes the spectrometer and creates the data streams and properties.
+
+        Raises
+        ------
+        RuntimeError
+            If the spectrometer cannot be found.
+        '''
+
+        # Define and set defaut exposure time
+        self.exposure_time = self._exposure_time
+
+        # Create datastreams
+        self.spectra = self.make_data_stream('spectra', 'float32', [self.pixels_number], self.NUM_FRAMES)
+
+        self.is_saturating = self.make_data_stream('is_saturating', 'int8', [1], self.NUM_FRAMES)
+        self.is_saturating.submit_data(np.array([0], dtype='int8'))
+
+        # Create properties
+        def make_property_helper(name, read_only=False):
+            if read_only:
+                self.make_property(name, lambda: getattr(self, name))
+            else:
+                self.make_property(name, lambda: getattr(self, name), lambda val: setattr(self, name, val))
+
+        make_property_helper('exposure_time')
+        make_property_helper('wavelengths', read_only=True)
+
+    def main(self):
+        '''
+        The main function of the service.
+
+        This function is called when the service is started.
+        '''
+        while not self.should_shut_down:
+            intensities = self.take_one_spectrum(self.pixels_number)
+            self.is_saturating.submit_data(np.array([0], dtype='int8'))
+            self.spectra.submit_data(np.array(intensities, dtype='float32'))
```

Comment:
I am really starting to nit-pick but since we're on another review (probs last): Could you invert these two lines? Submit the intensities first (since they're "measured" before), then submit the zero saturation? A blank line before the sleep would also be good.